File: 2197-replace-non-coprime-numbers-in-array/2197-replace-non-coprime-numbers-in-array.cpp
class Solution {
public:
    vector<int> replaceNonCoprimes(vector<int>& nums) {
        int n = nums.size();
        vector<int> rets;
        for(int i = 0; i < n; ++i) {
            // int x = nums[i];
            // while(!rets.empty() && gcd(rets.back(), x) > 1) {
            //     x = lcm(rets.back(), x);
            //     rets.pop_back();
            // }
            // rets.push_back(x);
                        int x = nums[i];
            while (!rets.empty() && gcd(rets.back(), x)>1)
            {
                x = lcm(rets.back(), x);
                rets.pop_back();                
            }
            rets.push_back(x);
        }
        
        return rets;
    }
    
    long long lcm(long long x, long long y) {
        return x * y / gcd(x, y);
    }
};

File: 2197-replace-non-coprime-numbers-in-array/2197-replace-non-coprime-numbers-in-array.py
class Solution:
    def replaceNonCoprimes(self, nums: List[int]) -> List[int]:
        n = len(nums)
        res = []
        
        def gcd(a, b):
            while b:
                a, b = b, a % b
            return a
        
        def lcm(x, y):
            return (x * y) // gcd(x, y)
        
        for x in nums:
            while res and gcd(res[-1], x) > 1:
                x = lcm(res[-1], x)
                res.pop()
            res.append(x)
        return res